guppylang/module.py

```diff
@@ -63,44 +70,95 @@ class GuppyModule:
     def __init__(self, name: str, import_builtins: bool = True):
         self.name = name
         self._globals = Globals({}, {}, {}, {})
-        self._compiled_globals = {}
         self._imported_globals = Globals.default()
-        self._imported_compiled_globals = {}
+        self._imported_checked_defs = {}
         self._compiled = False
         self._compiled_hugr = None
         self._instance_func_buffer = None
         self._raw_defs = {}
         self._raw_type_defs = {}
+        self._checked_defs = {}
 
         # Import builtin module
         if import_builtins:
             import guppylang.prelude.builtins as builtins
 
             self.load(builtins)
 
-    def load(self, m: Union[ModuleType, "GuppyModule"]) -> None:
-        """Imports another Guppy module."""
-        self._check_not_yet_compiled()
-        if isinstance(m, GuppyModule):
-            # Compile module if it isn't compiled yet
-            if not m.compiled:
-                m.compile()
-
-            # For now, we can only import custom functions
-            if any(
-                not isinstance(v, CustomFunctionDef | TypeDef | ParamDef)
-                for v in m._compiled_globals.values()
-            ):
-                raise GuppyError(
-                    "Importing modules with defined functions is not supported yet"
-                )
+    def load(
+        self,
+        *args: "Definition | GuppyModule | ModuleType",
+        **kwargs: "Definition | GuppyModule | ModuleType",
+    ) -> None:
+        """Imports another Guppy module or selected definitions from a module.
 
-            self._imported_globals |= m._globals
-            self._imported_compiled_globals |= m._compiled_globals
-        else:
-            for val in m.__dict__.values():
-                if isinstance(val, GuppyModule):
-                    self.load(val)
+        Keyword args may be used to specify alias names for the imports.
+        """
+        modules: set[GuppyModule] = set()
+        defs: dict[DefId, CheckedDef] = {}
+        names: dict[str, DefId] = {}
+
+        # Collect imports in reverse since we'll use it as a stack to push and pop from
+        imports: list[tuple[str, Definition | GuppyModule | ModuleType]] = [
+            *reversed(kwargs.items()),
+            *(("", arg) for arg in reversed(args)),
+        ]
+        while imports:
+            alias, imp = imports.pop()
+            if isinstance(imp, Definition):
```

Comment:
It'd be nice to check for name clashes. I.e. if `mod_a` and `mod_b` both define `foo`, and I write

```
mod_c = GuppyModule('')
mod_c.load(mod_a)
mod_c.load(mod_b)

@guppy(mod_c)
def bar() -> ():
    return foo()
```

Then I assume `mod_b`'s `foo` is shadowing the one from `mod_a`, but I'd rather this was made explicit

Reply:
Not sure, Python let's you shadow stuff if you do `from module import *`.

In the past we were for example checking that people don't define functions with clashing names, but this got annoying in the Jupyter setting. My guess is that the same applies here

Reply:
If that's the python semantics then I guess it's okay for this example. How does python deal with aliases? Because if we make an alias that could also be shadowed by something else, which seems a bit odd

Reply:
> Because if we make an alias that could also be shadowed by something else, which seems a bit odd

That's Python for you 😅

`import` brings names into scope, override whatever was bound to the name before

Reply:
At least it's consistent!